source/auth/auth.rst

```diff
@@ -1272,151 +1215,167 @@ mechanism_properties
-If the callback succeeds, the new Access Token MUST be sent using a JwtStepRequest.  If the request fails with a ReauthenticationRequired error, that error MUST be propagated to the user.
-The driver MUST have a guard or a flag in place to differentiate between a JwtStepRequest ReauthenticationRequired failure that takes place after a PrincipalStepRequest has been made to prevent an infinite loop.
+  interface RequestParameters {
+      // Timeout in seconds for the callback.  Optionally, timeoutContext instead if applicable to language.
+      timeoutSeconds: int;
 
-If there is no refresh callback and no unexpired Access Token, the request callback will be called.  Multithreaded drivers MUST ensure that there is at most one concurrent invocation of the above fallback logic for a given cache key.
+      // The version of the callback parameter interface.
+      version: int;
 
-If a cached value is used and the authentication step fails or times out, the driver MUST clear the
-cached value.
+      // The refresh token, if applicable, to be used by the callback to request a new token from the issuer.
+      refreshToken?: str;
+  }
 
+.. code: typescript
 
-Reauthentication
-````````````````
-When reauthentication is requested by the server (as a 391 error code) and MONGODB-OIDC is in use, the driver MUST
-ensure that the Access Token that was most-recently used to authenticate this connection is not used for subsequent authentication, by marking it as expired. If non-expired Access Token is available in the cache, it should be used as usual.  If a refresh
-callback is given, it will be called as usual.  Otherwise the IdPServerResponse will be cleared if present and authentication will proceed from the request callback.
+  function onRequest(info: IdpInfo, params: RequestParameters): IdpResponse
 
-If the ``sasl`` step(s) fail with a 391 error code and the payload of the command contained ``jwt`` , the driver MUST clear the IdPServerResponse and
-attempt to authenticate one more time starting from
-``PrincipalStepRequest``.  An initial reauthentication may fail for various reasons, such as token expiration or identity provider reconfiguration, so a second reauthentication might be needed.
+Before calling the callback, the driver MUST acquire a lock unique to the ``MongoClient``.
+The driver MUST ensure that credentials have not changed between when the lock was requested and when it was acquired.  The lock MUST be released
+when the callback call as finished or errored.
+This is because the ``REQUEST_TOKEN_CALLBACK`` may involve human interaction, and refresh tokens migh only be able to be used used once.
```

Comment:
```suggestion
This is because the ``REQUEST_TOKEN_CALLBACK`` may involve human interaction, and refresh tokens might only be able to be used used once.
```